src/requestManager.cpp

```diff
@@ -428,6 +428,51 @@ bool CUnknownObj::AddSource(CNode *from)
     return false;
 }
 
+void CRequestManager::AddRequestedHash(const uint256 blockHash)
+{
+    LOCK(cs_requestedHashes);
+    requestedHashes.insert(blockHash);
+}
+
+bool CRequestManager::CheckRequestedHash(const uint256 blockHash)
+{
+    LOCK(cs_requestedHashes);
+    return requestedHashes.count(blockHash);
+}
+
+void CRequestManager::RemoveRequestedHash(const uint256 blockHash)
+{
+    LOCK(cs_requestedHashes);
+    requestedHashes.erase(blockHash);
+}
+
+bool CRequestManager::RequestBlockByHash(const uint256 &blockHash)
+{
+    if (CheckRequestedHash(blockHash))
+        return true;
+    // set it to MSG_BLOCK here but it should get overwritten in RequestBlock
+    CInv obj(2, blockHash);
+    CNode *pfrom;
+    {
+        LOCK(cs_vNodes);
+        for (CNode *pnode : vNodes)
+        {
+            // peer is a fullnode
+            if (pfrom->nServices & NODE_NETWORK)
+            {
+                pfrom = pfrom;
+                break;
+            }
+        }
+    }
+    bool requested = RequestBlock(pfrom, obj);
```

Comment:
should check that pfrom is not null here in case you didn't get a node_network peer.